test/integration/integration_test.cc

```diff
@@ -2427,6 +2427,97 @@ TEST_P(IntegrationTest, RandomPreconnect) {
   }
 }
 
+/*
+This test ensures that the feature flag: envoy.restart_features.ensure_connection_retry
+enables retries on connection timeouts up to the limit per route instead of the limit
+coming from the downstream listener.
+*/
+TEST_P(IntegrationTest, EnsureConnectionRetry) {
+  // Set buffer limit upstream and downstream.
+  config_helper_.setBufferLimits(100, 64);
+  // Increase the base backoff to ensure we have time to see the connection failure and start
+  // the new upstream before the retry.
+  config_helper_.addRuntimeOverride("upstream.base_retry_backoff_ms", "200");
+  // enable feature flags envoy.restart_features.ensure_connection_retry
+  config_helper_.addRuntimeOverride("envoy.restart_features.ensure_connection_retry", "true");
```

Comment:
Could you also add a test to verify other retry features together with connection retries?

Reply:
Done, I added some test based on 5xx retry with the feature enable:
1. test that we still retry 5xx when body is smaller than limit
2. test that we don't retry 5xx when body is bigger than limit